docs/source/architecture.rst

```diff
@@ -4,14 +4,62 @@
+Or it could be that we wish to systematically alter the notebook, for example by clearing all output cells.
+Format agnostic operations on cell content that do not violate the nbformat spec can be interpreted as a notebook to notebook conversion step; such operations can be performed as part of the preprocessing step.
+
+But often we want to have the notebook's structured content in a different format.
+Importantly, in many cases the structure of the notebook should be reflected in the structure of the output, adapted to the output's format.
+For that purpose, the original JSON structure of the document is crucial scaffolding needed to support this kind of structured output.
+In order to maintain structured, it can be useful to apply our conversion programmatically on the structure itself.
+To do so, when converting to formats other than the notebook, we use the `jinja`_ templating engine.
+
+The basic unit of structure in a notebook is the cell.
+Accordingly, since our templating engine is capable of expressing structure, the basic unit in our templates will often be specified at the cell level.
+Each cell has a certain type; the three most important cell types for our purposes are code, markdown, and raw NbConvert.
+Code cells can be split further into their input and their output.
+Operations can also occur separately on input and output and their respective subcomponents.
+Markdown cells and raw NbConvert cells do not have analogous substructure.
+
+The template's structure then can be seen as a mechanism for selecting content on which to operate.
+Because the template operates on individual cells, this has some upsides and drawbacks.
+One upside is that this allows the template to have access to the individual cell's metadata, which enables intelligently transforming the appropriate content. 
+The transformations occur as a series of replacement rules and filters. 
+For many purposes these filters take the form of external calls to `pandoc`_, which is a utility for converting between many different document formats.
+One downside is that this makes operations that require global coördination (e.g., cross referencing across cells) somewhat challenging to implement as filters inside templates.
+
+Note that all that we've described is happening in memory. 
+This is crucial in order to ensure that this functionality is available when writing files is more challenging.
+Nonetheless, the reason for using nbconvert almost always involves producing some kind of output file.
+We take the in-memory object and write a file appropriate for the output type.
+
+The entirety of heretofore described process can be described as part of an ``Exporter``. 
+``Exporter``s often involves ``Preprocessor``s, ``filters``, ``templates`` and ``Writer``s. 
+These classes and functions are described in greater detail below.
```

Comment:
I'm still not 100% sure about way that the resource dictionary interacts with everything else. We should chat about that more.